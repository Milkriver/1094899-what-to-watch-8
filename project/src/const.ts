export enum AppRoute {
  Main = '/',
  SignIn = '/login',
  MyList = '/mylist',
  Film = '/films/:id',
  AddReview = '/films/:id/review',
  Player = '/player/:id'
}

export enum AuthorizationStatus {
  Auth = 'AUTH',
  NoAuth = 'NO_AUTH',
  Unknown = 'UNKNOWN',
}

export enum APIRoute {
  Movies = '/films',
  SingleMovie = '/films/:id',
  SameGenreMovies = '/films/:id/similar',
  PromoMovie = '/promo',
  FavouriteMovies = '/favorite',
  StatusFavouriteMovie = '/favorite/: film_id/: status',
  Reviews = '/comments/:id',
  Login = '/login',
  Logout = '/logout',
}
