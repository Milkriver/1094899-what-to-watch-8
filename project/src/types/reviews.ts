export interface IReview {
    id: number,
    movieId: number,
    author: string,
    date: string,
    rating: number,
    text: string,
  }
