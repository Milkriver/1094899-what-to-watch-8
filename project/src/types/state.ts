import { AuthorizationStatus } from '../const';
import { IMovie } from './common';
import { IReviewResponse } from './reviews';

export interface IState {
  currentGenre: string
  movies: IMovie[]
  isDataLoaded: boolean
  authorizationStatus: AuthorizationStatus,
  movie: IMovie,
  sameMovies: IMovie[],
  reviews: IReviewResponse[],
  promoMovie: IMovie,
}
