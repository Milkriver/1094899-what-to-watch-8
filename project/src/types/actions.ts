import { AxiosInstance } from 'axios';
import { Action } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';
import { IState } from './state';

export enum ActionType {
  ChangeGenre = 'movieList/ChangeGenre',
  LoadMovies = 'movieList/LoadMovies',
  LoadSingleMovie = 'movieCard/LoadSingleMovie',
  LoadActiveMovie = 'movieCard/LoadActiveMovie',
  LoadSameGenreMovies = 'movieCard/LoadSameGenreMovies',
  LoadPromoMovie = 'movieCard/LoadPromoMovie',
  LoadFavouriteMovies = 'userList/LoadFavouriteMovies',
  AddFavoriteMovie = 'userList/AddFavoriteMovie',
  RemoveFavoriteMovie = 'userList/RemoveFavoriteMovie',
  LoadReviews = 'movieCard/LoadReviews',
  AddReview = 'movieCard/AddReview',
  RequireAuthorization = 'user/requireAuthorization',
  RequireLogout = 'user/requireLogout',
  LoadUserData = 'user/loadUserData',
}

export type ThunkActionResult<R = Promise<void>> = ThunkAction<R, IState, AxiosInstance, Action>;

export type ThunkAppDispatch = ThunkDispatch<IState, AxiosInstance, Action>;
