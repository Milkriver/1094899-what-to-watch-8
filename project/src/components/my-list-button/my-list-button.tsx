import { useEffect, useState } from 'react';
import { connect, ConnectedProps, useDispatch } from 'react-redux';
import { useHistory } from 'react-router';
import { AppRoute, AuthorizationStatus, FavoriteMoviesStatus } from '../../const';
import { changeFavoriteMovies, fetchFavoriteMovies } from '../../store/api-actions';
import { ThunkAppDispatch } from '../../types/actions';
import { IMovie } from '../../types/common';
import { IState } from '../../types/state';

type IProps = {
  currentMovie: IMovie,
}


const mapStateToProps = ({ authorizationStatus }: IState) => ({
  authorizationStatus,
});

const mapDispatchToProps = (dispatch: ThunkAppDispatch) => ({
  onChangeFavoriteMovies(cardId: number, status: FavoriteMoviesStatus) {
    dispatch(changeFavoriteMovies(cardId, status));
    dispatch(fetchFavoriteMovies());
  },
});

const connector = connect(mapStateToProps, mapDispatchToProps);
type PropsFromRedux = ConnectedProps<typeof connector>;
type ConnectedComponentProps = PropsFromRedux & IProps;

function MyListButton({ currentMovie, authorizationStatus }: ConnectedComponentProps): JSX.Element {
  const [favouriteMovieStatus, setFavouriteMovieStatus] = useState(currentMovie.isFavorite);
  const currentMovieId = currentMovie.id;
  const history = useHistory();
  const dispatch = useDispatch();
  useEffect(() => setFavouriteMovieStatus(currentMovie?.isFavorite), [currentMovie]);
  const onHandleFavoriteMovieClick = () => {
    if (authorizationStatus === AuthorizationStatus.Auth) {
      dispatch(changeFavoriteMovies(currentMovieId, favouriteMovieStatus ? FavoriteMoviesStatus.Remove : FavoriteMoviesStatus.Add));
      setFavouriteMovieStatus(!favouriteMovieStatus);
      history.push(AppRoute.MyList);
    } else {
      history.push(AppRoute.SignIn);
    }
  };

  return (
    <button
      className="btn btn--list film-card__button"
      type="button"
      onClick={onHandleFavoriteMovieClick}
    >
      <svg viewBox="0 0 19 20" width="19" height="20">
        <use xlinkHref={favouriteMovieStatus ? '#in-list' : '#add'}></use>
      </svg>
      <span>My list</span>
    </button>
  );
}
export { MyListButton };
export default connector(MyListButton);
