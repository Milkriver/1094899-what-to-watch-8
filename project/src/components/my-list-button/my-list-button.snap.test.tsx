import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { IMovie } from '../../types/common';
import { configureMockStore } from '@jedmao/redux-mock-store';
import { MyListButton } from './my-list-button';
import { AuthorizationStatus } from '../../const';


const mockMovie1: IMovie = {
  id: 1, name: 'name1', posterImage: 'posterImage1', previewImage: 'previewImage1', backgroundImage: 'backgroundImage1',
  backgroundColor: 'black', videoLink: 'videoLink1', previewVideoLink: 'previewVideoLink1',
  description: 'description1', rating: 1, scoresCount: 1, director: 'director1',
  starring: ['actor1', 'actor2'], runTime: 1, genre: 'Comedy', released: 1, isFavorite: true,
};
const mockFn = () => Promise.resolve();

describe('Component: MyListButton', () => {
  const mockStore = configureMockStore();
  const store = mockStore({
    MOVIES: {},
    USER: {},
    REVIEWS: {},
  });

  store.dispatch = jest.fn();

  test('should render correctly', () => {
    const page = (
      <Provider store={store}>
        <MyListButton
          currentMovie={mockMovie1}
          authorizationStatus={AuthorizationStatus.Auth}
          onChangeFavoriteMovies={mockFn}
        />
      </Provider>
    );
    const { container } = render(page);
    expect(container).toMatchSnapshot();
  });
});
