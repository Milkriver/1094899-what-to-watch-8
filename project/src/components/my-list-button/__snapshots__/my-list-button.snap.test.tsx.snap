// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: MyListButton should render correctly 1`] = `
<div>
  <button
    class="btn btn--list film-card__button"
    type="button"
  >
    <svg
      height="20"
      viewBox="0 0 19 20"
      width="19"
    >
      <use
        xlink:href="#in-list"
      />
    </svg>
    <span>
      My list
    </span>
  </button>
</div>
`;
